library/Faktory/Worker.hs

```diff
@@ -86,87 +89,82 @@ untilM_ predicate action = do
         untilM_ predicate action
     )
 
--- | Creates a new faktory worker, @'action'@ is ran with @'WorkerConfig'@ before
--- polling begins. Jobs received are passed to @'handler'@. The worker's
--- connection is closed when job processing ends.
-withRunWorker ::
-  (HasCallStack, FromJSON args)
-  =>  Settings
+-- | Returns the @'WorkerId' associated with a @'Worker'.
+workerID :: Worker -> WorkerId
+workerID Worker{workerId} = workerId
+
+-- | Forks a new faktory worker, continuously polls the faktory server for
+-- jobs which are passed to @'handler'@. The client is closed when the forked
+-- thread ends.
+startWorker
+  :: (HasCallStack, FromJSON args)
+  => Settings
   -> WorkerSettings
-  -> (WorkerConfig -> IO a)
   -> (Job args -> IO ())
-  -> IO ()
-withRunWorker settings workerSettings action handler =
-  configureWorker settings workerSettings
-    $ \config -> do
-        void $ action config
-        runWorkerWithConfig handler config
+  -> IO (ThreadId, Worker)
+startWorker settings workerSettings handler = do
+  workerId <- maybe randomWorkerId pure $ settingsId workerSettings
+  isQuieted <- newTVarIO False
+  client <- newClient settings $ Just workerId
+  isDone <- newEmptyMVar
+  let worker = Worker{client, isDone, isQuieted, settings, workerId, workerSettings}
+  tid <-
+    forkFinally
+      ( do
+          beatThreadId <- forkIOWithThrowToParent $ forever $ heartBeat worker
+          finally
+            ( flip runReaderT worker . runWorkerM $
+                untilM_ shouldStopWorker (processorLoop handler)
+                  `catch` (\(_ex :: WorkerHalt) -> pure ())
+            )
+            $ killThread beatThreadId
+      )
+      ( \e -> do
+          closeClient client
+          case e of
+            Left err -> throwIO err
```

Comment:
Do we need to also `putMVar isDone ()` here? What is the state of the worker after this occurs? Can the worker be reused? What happens if someone catches this and then tries to quiet, wait on, or start this worker again?